cellarium/cas/_utils.py

```diff
@@ -0,0 +1,67 @@
+"""
+This module provides utility functions to handle matrix files and calculate optimal chunk sizes based on memory
+constraints.
+
+Constants:
+-----------
+RESERVE_COEFFICIENT: float
+    A coefficient used to define a chunk size. This number helps to keep the chunk size below the provided limit
+"""
+import math
+import os
+import tempfile
+import typing as t
+
+import anndata
+
+RESERVE_COEFFICIENT = 0.9
+
+
+def round_down_to_nearest(x: t.Union[int, float], nearest: int) -> int:
+    """
+    Rounds down a number to the nearest specified number.
+
+    :param x: The number to be rounded down.
+    :param nearest: The number to which `x` should be rounded down to.
+    """
+    return math.floor(x / nearest) * nearest
+
+
+def get_preferred_chunk_size(
```

Comment:
This implementation is fine for most cases, but consider a case where the adata is a concatenation of two datasets, one is very dense and the second is very sparse. The uniform approximation fails in this case.

I still think the idea of dynamically generating chunks is safer. We can use your preferred chunk size calculation to initialize a good default chunk size. Here's an implementation (NOT TESTED, PLEASE REVIEW AND TEST):

```
def generate_adata_chunks(
        adata: AnnData,
        shrink_factor: float = 0.75,
        growth_factor: float = 1.25,
        max_shrink_attempts: int = np.inf,
        max_growth_attempts: int = np.inf,
        chunk_size_limit_in_mb: float) -> t.Generator[AnnData, None, None]:
    total_cells = len(adata)
    current_pos = 0
    delta = get_preferred_chunk_size(adata, chunk_size_limit_in_mb)
    while current_pos < total_cells:
        # shrink chunk size if needed
        n_shrink_attempts = 0
        chunk_size_mb = np.inf
        while chunk_size_mb > chunk_size_limit_in_mb:
            assert n_shrink_attempts < max_shrink_attempts, "Too many adata chunk size shrink attempts!"
            delta = int(delta * shrink_factor)
            assert delta > 0, "Cannot generate an anndata chunk within the specified limit ({chunk_size_limit_in_mb:.1f}mb)"
            end_pos = min(current_pos + delta, total_cells)
            chunk_size_mb = calculate_chunk_size(adata[current_pos:end_pos])
            n_shrink_attempts += 1
        # grow chunk size if possible
        n_growth_attempts = 0
        while n_growth_attempts < max_growth_attempts:
            enlarged_delta = int(delta * growth_factor)
            enlarged_end_pos = min(current_pos + enlarged_delta, total_cells)
            chunk_size_mb = calculate_chunk_size(adata[current_pos:enlarged_end_pos])
            if chunk_size_mb < chunk_size_limit_in_mb:
                n_growth_attempts += 1
                delta = enlarged_delta
                end_pos = enlarged_end_pos
            else:
                break
        yield adata[current_pos:end_pos]
        current_pos = end_pos
```